### PR TITLE
[FIX] base: speedup access rights check query

### PR DESCRIPTION
**Pull request by Xavier-Do**
Since #84736 the access right check queries were merged into one. 
Unfortunately the query plan for cases of a group with tons of users can be suboptimal.

Using a subselect will force the query plan in order to avoid an index only scan on res_groups_users_rel.

Performances where tested in production, this new query being around 300 and 1000 times faster for two tests cases (res.partner
 and account.payment.term)

### CODE REVIEW

**Review by VincentSchippefilt — APPROVED** [review]

**Comment by odony** [comment]
Thanks! As an alternative that is maybe slightly more readable, here's a variation of the original query, with the same performance boost.

By forcing the `uid` clause **in the JOIN condition**, we ensure that PG can efficiently retrieve m2m rows based on the `uid` index (low cardinality per leaf, since it's bound by the number of groups).

And then we keep only matches, excluding the ACL entries that had a `group_id` but did not match the JOIN condition because the user is not a member of those groups.


```patch
diff --git odoo/addons/base/models/ir_model.py odoo/addons/base/models/ir_model.py
index ae4229011c7..500ed6dd3d4 100644
--- odoo/addons/base/models/ir_model.py
+++ odoo/addons/base/models/ir_model.py
@@ -1769,11 +1769,11 @@ class IrModelAccess(models.Model):
         self._cr.execute("""SELECT BOOL_OR(perm_{mode})
                               FROM ir_model_access a
                               JOIN ir_model m ON (m.id = a.model_id)
-                         LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
+                         LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id AND gu.uid = %s)
                              WHERE m.model = %s
-                               AND (a.group_id IS NULL OR gu.uid = %s)
+                               AND (a.group_id IS NULL OR a.group_id = gu.gid)
                                AND a.active IS TRUE""".format(mode=mode),
-                         (model, self._uid,))
+                         (self._uid, model,))
         r = self._cr.fetchone()[0]
 
         if not r and raise_exception:
```

**Comment by VincentSchippefilt** [comment]
@odony Did you re-run the both query in test ? Does postgres actually use the correct indexes ?

File: odoo/addons/base/models/ir_model.py [hunk 1]
```diff
@@ -1769,10 +1769,17 @@ def check(self, model, mode='read', raise_exception=True):
...
                               FROM ir_model_access a
                               JOIN ir_model m ON (m.id = a.model_id)
-                         LEFT JOIN res_groups_users_rel gu ON (gu.gid = a.group_id)
                              WHERE m.model = %s
-                               AND (a.group_id IS NULL OR gu.uid = %s)
-                               AND a.active IS TRUE""".format(mode=mode),
+                               AND a.active IS TRUE
+                               AND (
+                                    a.group_id IS NULL OR
+                                    a.group_id in (
```
Comment (rco-odoo) [inline odoo/addons/base/models/ir_model.py:1776]:
```suggestion
                                    a.group_id IN (
```

**Comment by rco-odoo** [comment]
> Thanks! As an alternative that is maybe slightly more readable, here's a variation of the original query, with the same performance boost.

Which alternative is slightly more readable?  Your proposal is way more obscure to me, and the difference is _very_ subtle.  The subselect is crystal clear to me, and the naive query plan is the best plan.

**Review by rco-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by Xavier-Do** [comment]
> That's what gives the planner the most options for selecting an appropriate query plan, depending on statistics: nested loop, merge join, hash join, etc.

The idea here is that the planner doesn't choose the right query plan, using the sub select is to help him to make a better choice so it looks ok for me to restraint the possibles query plan. We know that we shouldn't have thousands of groups. 

This is also easier to understand IMHO. User has access if the ir_model_access has no group or corresponds to one of the user groups. With the join version I need some time to understand and convince myself that it is correct especially because of the double `a.group_id = gu.gid` condition only needed to manage the null case of the left_join.

Actually I'm not actually convinced that the `AND (a.group_id IS NULL OR a.group_id = gu.gid)` is still usefull in your version since all line should either have an empty group (LEFT JOIN) or a.group_id = gu.gid (ON (gu.gid = a.group_id AND gu.uid = %s)

Since both queries are ok for you we will stay on the initial version.

**Comment by Xavier-Do** [comment]
> Since both queries are ok for you we will stay on the initial version.

**Comment by Xavier-Do** [comment]
Removed the review because of this missclick...

@robodoo r+

**Comment by odony** [comment]
> Actually I'm not actually convinced that the `AND (a.group_id IS NULL OR a.group_id = gu.gid)` is still usefull in your version since all line should either have an empty group (`LEFT JOIN`) or have corresponding groups (`ON (gu.gid = a.group_id AND gu.uid = %s`)

You still need to exclude ACL rows that do have a `group_id` but didn't match the LEFT JOIN because the user is not a member of that group. That's where `gid = group_id` comes into play, as those lines will have a NULL `gid`.

**Comment by rco-odoo** [comment]
> > Which alternative is slightly more readable?
> 
> I guess that depends on your database habits ;-) In RDBMS terms, when you're querying two tables linked by a FK, you're naturally inclined to use a JOIN variant, not a sub-query. That's what gives the planner the most options for selecting an appropriate query plan, depending on statistics: nested loop, merge join, hash join, etc. With an explicit subquery that's not true, and the planner can't second-guess your choice.

You are right, I am certainly not as good as you are in SQL.  I really don't like those LEFT JOINS with OR conditions in the WHERE clause, because it requires more non-trivial reasoning with NULLs.  The subselect is much easier to understand to me.

**Pull request closed without merging**